A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,425 @@
+- __Route Discovery Service (RDS)__: Returns `RouteConfiguration`
+  resources.  Provides data used to populate the gRPC service config.  Points
+  to the Cluster.
+- __Cluster Discovery Service (CDS)__: Returns `Cluster` resources.
+  Configures things like load balancing policy and load reporting.  Points
+  to the ClusterLoadAssignment.
+- __Endpoint Discovery Service (EDS)__: Returns `ClusterLoadAssignment`
+  resources.  Configures the set of endpoints (backend servers) to load
+  balance across and may tell the client to drop requests.
+
+gRPC will support the Aggregate Discovery Service (ADS) variant of xDS,
+where all of these resource types are obtained on a single gRPC stream.
+However, the different phases of the API flow are still typically
+referred to using the separate service names described above.
+
+### Related Proposals: 
+
+[A24: Load Balancing Policy Configuration](https://github.com/grpc/proposal/blob/master/A24-lb-policy-config.md)
+
+## Proposal
+
+### gRPC Client Architecture
+
+Because xDS handles not only load balancing but also service discovery
+and configuration, gRPC will support xDS via both the resolver and LB
+policy plugins.  Both the resolver and LB policy plugins will need to
+communicate with the xDS server, so the functionality for interacting with
+the xDS server will be contained within an XdsClient object, which will
+be used by both the resolver and the LB policies.
+
+There will be two separate LB policies for xDS, one to support `Cluster`
+data, and another to support `ClusterLoadAssignment` data.
+
+Note that both the resolvers and the LB policies will have
+"experimental" suffixes for their names for now.  These suffixes will be
+removed when the functionality has proven to be stable.
+
+![gRPC Client Architecture Diagram](A27_graphics/grpc_client_architecture.png)
+
+[Link to SVG file](A27_graphics/grpc_client_architecture.svg)
+
+#### XdsClient and Bootstrap File
+
+The XdsClient object contains all of the logic for interacting with the
+xDS server.  The XdsClient object is a purely internal API, not
+something exposed to gRPC users, but it is described here as part of
+the architecture used to support xDS in the gRPC client.
+
+The XdsClient object is configured via a bootstrap file.  The location
+of the bootstrap file is determined via the `GRPC_XDS_BOOTSTRAP`
+environment variable.  The file is in JSON form, and its contents look
+like this:
+
+```
+{
+  // The xDS server to talk to.  The value is an array to allow for a
+  // future change to add support for failing over to a secondary xDS server
+  // if the primary is down, but for now, only the first entry in the
+  // array will be used.
+  xds_servers": [
+    {
+      "server_uri": <string containing URI of xds server>,
+      // List of channel creds; client will stop at the first type it
+      // supports.  This field is optional; if not specified, xDS will use
+      // the same channel creds as the backends, but with any associated
```

Comment:
Same creds for backends and xDS server seems to be a very specialized use-case that's not going to be common (I think). Instead of making that the default behavior how about a "channel_creds" value - something like "same-as-backend" here?

Reply:
This may not be super common, but it won't be unheard of either.  We have some internal use-cases that need this behavior.

If a user does not want this default behavior in their deployment, then they can just treat this as a required field.

For the security work (not in scope for this doc), our plan is to make a registry of credential types, much like we have for resolvers and LB policies.  The name used here would be the name to look up in that registry.  I don't think it makes sense to force a magic value into that registry.

Reply:
Okay.